src/m2n/PointToPointCommunication.cpp

```diff
@@ -245,42 +281,65 @@ void printLocalIndexCountStats(std::map<int, std::vector<int>> const &m)
+    mesh::Mesh::VertexDistribution &otherVertexDistribution,
+    int                             thisRank = utils::IntraComm::getRank())
 {
   auto iterator = thisVertexDistribution.find(thisRank);
   if (iterator == thisVertexDistribution.end())
     return {};
 
-  // Build lookup table from otherIndex -> rank for the otherVertexDistribution
-  const auto lookupIndexRank = [&otherVertexDistribution] {
-    boost::container::flat_multimap<int, int> lookupIndexRank;
-    for (const auto &other : otherVertexDistribution) {
-      for (const auto &otherIndex : other.second) {
-        lookupIndexRank.emplace(otherIndex, other.first);
-      }
-    }
-    return lookupIndexRank;
-  }();
+  std::map<int, std::vector<int>> communicationMap;
+
+  // take advantage that these data structures are in most cases sorted by construction,
+  // i.e., we perform here mostly a safety check and don't perform an actual sorting
+  if (!std::is_sorted(iterator->second.begin(), iterator->second.end())) {
+    std::sort(iterator->second.begin(), iterator->second.end());
+  }
 
-  auto const &indices = iterator->second;
+  // now we iterate over all other vertex distributions to compute the intersection
+  for (auto &other : otherVertexDistribution) {
+    // first a safety check, that we are actually sorted, similar to above
+    if (!std::is_sorted(other.second.begin(), other.second.end())) {
+      std::sort(other.second.begin(), other.second.end());
+    }
 
-  std::map<int, std::vector<int>> communicationMap;
-  for (size_t index = 0lu; index < indices.size(); ++index) {
-    auto range = lookupIndexRank.equal_range(indices[index]);
-    for (auto iter = range.first; iter != range.second; ++iter) {
-      communicationMap[iter->second].push_back(index);
+    // before starting to compute an actual intersection, we first check if elements can
+    // possibly be in both data sets by comparing upper and lower index bounds of both
+    // data sets. For typical partitioning schemes, each rank only exchanges data with
+    // a few neighbors such that this check already filters out a significant amount of
+    // computations
+    if (iterator->second.empty() || other.second.empty() || (other.second.back() < iterator->second.at(0)) || (other.second.at(0) > iterator->second.back())) {
+      // in this case there is nothing to be done
+      continue;
+    } else {
+      // we have an intersection, let's compute it
+      std::vector<int> inters;
+      // the actual worker function, which gives us the indices of intersecting elements
+      // have a look at the documentation of the function for more details
+      precice::m2n::set_intersection_indices(iterator->second.begin(), iterator->second.begin(), iterator->second.end(),
+                                             other.second.begin(), other.second.end(),
+                                             std::back_inserter(inters));
```

Comment:
I think that this will be the biggest bottleneck after your changes.

Suggestion:
We move the `inters` outside the for loop and `clear()` it here instead.
The capacity of the vector will not shrink by doing this, which essentially is a form of a preallocation heuristic.

The below, you insert a copy of `inters` into the communication map.
This will copy the content and result in a vector with `capacity() == size()`, which should be more memory efficient.

Reply:
> I think that this will be the biggest bottleneck after your changes.

Well, it is the function doing the work, so in principle yes, but the function as a whole is now part of the bottle body rather than the neck. There are other functions which are now much more relevant in terms of performance and I exceeded my performance targets already by a lot and don't want to spend more effort. Still, the proposed changes are easy to test (ignoring the

> could impact further processing of the `communicationMap`

which I cannot really test, I think a different data type for the `communicationMap` should be the way to go anyway).

The proposed version is in fact slightly slower than the current version on both, my local machine as well as on SuperMUC-NG (tested the 240 cores version). On my local machine, we have around 306 us on the primary rank and 267 us on the most expensive secondary rank (the mesh is slightly smaller), using your proposed changes we end up with 434us on the primary rank and 283 us on the most expensive secondary rank.
On SuperMUC-NG we have around 2.200 us for the ccurrent version (as stated in the initial comment) and around 2.480 us for the proposed changes.

However, the scale on which these changes are observable is actually negligible, which brings me back to the first point.

Reply:
Thanks for testing! The results really surprise me. However, it's not worth investing more time here for such minuscule gains.